File: schemas/multi-language-i18n/convex/import_export_jobs.ts
// import_export_jobs: Bulk import/export operations.
// See README.md for full design rationale.

import { defineTable } from "convex/server";
import { v } from "convex/values";

export const import_export_jobs = defineTable({
  type: v.union(v.literal("import"), v.literal("export")),
  format: v.string(),
  status: v.union(
    v.literal("pending"),
    v.literal("processing"),
    v.literal("completed"),
    v.literal("failed")
  ),
  localeId: v.optional(v.id("locales")),
  namespaceId: v.optional(v.id("namespaces")),
  filePath: v.optional(v.string()),
  totalCount: v.number(), // Default 0.
  processedCount: v.number(), // Default 0.
  errorMessage: v.optional(v.string()),
  options: v.optional(v.any()),
  createdBy: v.optional(v.id("users")),
  startedAt: v.optional(v.number()),
  completedAt: v.optional(v.number()),
  updatedAt: v.number(),
})
  .index("by_status", ["status"])
  .index("by_created_by", ["createdBy"])
  .index("by_type_and_status", ["type", "status"]);

File: schemas/multi-language-i18n/spacetimedb/import_export_jobs.rs
// import_export_jobs: Bulk import/export job tracking.
// See README.md for full design rationale.

use spacetimedb::Timestamp;

/// Import = importing translations into the system.
/// Export = exporting translations from the system.
// type: String
#[derive(SpacetimeType, Clone)]
pub enum ImportExportType {
    Import,
    Export,
}

/// Pending = job is queued.
/// Processing = job is currently running.
/// Completed = job finished successfully.
/// Failed = job encountered an error.
// type: String
#[derive(SpacetimeType, Clone)]
pub enum ImportExportStatus {
    Pending,
    Processing,
    Completed,
    Failed,
}

// Composite index: (job_type, status) — all running import/export jobs by type

#[spacetimedb::table(name = import_export_jobs, public)]
pub struct ImportExportJob {
    #[primary_key]
    pub id: String, // UUID

    pub job_type: ImportExportType,
    pub format: String,

    #[index(btree)]
    pub status: ImportExportStatus,

    pub locale_id: Option<String>, // FK → locales.id (set null)
    pub namespace_id: Option<String>, // FK → namespaces.id (set null)
    pub file_path: Option<String>,
    pub total_count: i32,
    pub processed_count: i32,
    pub error_message: Option<String>,
    pub options: Option<String>, // JSON

    #[index(btree)]
    pub created_by: Option<String>, // FK → users.id (set null)

    pub started_at: Option<Timestamp>,
    pub completed_at: Option<Timestamp>,
    pub created_at: Timestamp,
    pub updated_at: Timestamp,
}
